src/client/sandbox/event/simulator.js

```diff
@@ -1,520 +1,483 @@
-            }
-        });
+    _dispatchKeyEvent (el, args) {
+        var ev = null;
+
+        if (document.createEvent) {
+            ev = document.createEvent('Events');
+            ev.initEvent(args.type, args.canBubble, args.cancelable);
+            ev = extend(ev, {
+                view:     args.view,
+                detail:   args.detail,
+                ctrlKey:  args.ctrlKey,
+                altKey:   args.altKey,
+                shiftKey: args.shiftKey,
+                metaKey:  args.metaKey,
+                keyCode:  args.keyCode,
+                charCode: args.charCode,
+                which:    args.which
+            });
 
-        raiseDispatchEvent(el, pointEvent, pArgs);
+            return this._raiseDispatchEvent(el, ev, args);
+        }
     }
 
-    ev = document.createEvent('MouseEvents');
-    ev.initMouseEvent(args.type, args.canBubble, args.cancelable, window, args.detail, args.screenX,
-        args.screenY, args.clientX, args.clientY, args.ctrlKey, args.altKey, args.shiftKey, args.metaKey,
-        args.button, args.relatedTarget);
+    _dispatchMouseEvent (el, args) {
+        var ev            = null;
+        var pointerRegExp = /mouse(down|up|move|over|out)/;
 
-    if (Browser.isMozilla || Browser.isIE) {
-        Object.defineProperty(ev, 'buttons', {
-            get: function () {
-                return args.buttons;
-            }
-        });
-    }
+        //NOTE: in IE submit doesn't work if a click is emulated for some submit button's children (for example img, B236676)
+        //In addition, if a test is being recorded in IE, the target of a click event is always a button, not a child, so child does not receive click event at all
+        if (Browser.isIE) {
+            if (args.type === 'click' || args.type === 'mouseup' || args.type === 'mousedown') {
+                if (el.parentNode && DOM.closest(el.parentNode, 'button')) {
+                    var closestButton = DOM.closest(el.parentNode, 'button');
 
-    //T188166 - act.hover trigger "mouseenter" event with "which" parameter 1
-    if (typeof args.which !== 'undefined' && Browser.isWebKit) {
-        Object.defineProperty(ev, Const.EVENT_SANDBOX_WHICH_PROPERTY_WRAPPER, {
-            get: function () {
-                return args.which;
+                    if (NativeMethods.getAttribute.call(closestButton, 'type') === 'submit')
+                        el = closestButton;
+                }
             }
-        });
-    }
+        }
 
-    //NOTE: After the MouseEvent was created by using initMouseEvent method pageX and pageY properties equal zero (only in IE9).
-    //We can set them only by defineProperty method (B253930)
-    if (Browser.isIE9) {
-        var currentDocument = DOM.findDocument(el);
-        var documentScroll  = Style.getElementScroll(currentDocument);
+        if (pointerRegExp.test(args.type) && (window.PointerEvent || window.MSPointerEvent)) {
+            var pointEvent       = Browser.version >
```

Comment:
This browser version checks bothers me. I don't see any browser type checks above, so seems like we check **any** browser for the version. Or we have IE check in the upper condition, but it's so unobvious that I was not able to parse it. Anyway, we need to refactor this code. \cc @helen-dikareva I guess